server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -128,6 +131,7 @@ protected void setNextReader(LeafReaderContext ctx, int[] docsInLeaf) throws IOE
                 this.leafNestedDocuments = nestedDocuments.getLeafNestedDocuments(ctx);
                 this.leafStoredFieldLoader = storedFieldLoader.getLoader(ctx, docsInLeaf);
                 this.leafSourceLoader = sourceLoader.leaf(ctx.reader(), docsInLeaf);
+                this.leafIdLoader = idLoader.leaf(leafStoredFieldLoader, ctx.reader(), docsInLeaf);
                 fieldLookupProvider.setNextReader(ctx);
```

Comment:
We probably need to wrap all these up into a SearchHitBuilder or something at some point, it's beginning to get a bit messy!